src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -3661,6 +3661,915 @@ class StubGenerator: public StubCodeGenerator {
+    //    //     v16[t+4] <- v16[t], v16[t+2], v14[t]
+    //    //  Input
+    //    //   v16 = {a[t],b[t],e[t],f[t]}       " = v16[t] "
+    //    //       = {h[t+2],g[t+2],d[t+2],c[t+2]}   " = v17[t+2] "
+    //    //   v17 = {f[t+2],e[t+2],b[t+2],a[t+2]}   " = v16[t+2] "
+    //    //   v14 = {W[t+3]+K[t+3], W[t+2]+K[t+2], W[t+1]+K[t+1], W[t+0]+K[t+0]}
+    //    //  Output:
+    //    //   v16 = {f[t+4],e[t+4],b[t+4],a[t+4]}   " = v16[t+4] "
+    //    vsha2ch.vv v16, v17, v14
+    //
+    //    // Combine 2QW into 1QW
+    //    //
+    //    // To generate the next 4 words, "new_v10"/"v14" from v10-v13, vsha2ms needs
+    //    //     v10[0..3], v11[0], v12[1..3], v13[0, 2..3]
+    //    // and it can only take 3 vectors as inputs. Hence we need to combine
+    //    // v11[0] and v12[1..3] in a single vector.
+    //    //
+    //    // vmerge Vt4, Vt1, Vt2, V0
+    //    // Input
+    //    //  V0 = mask // first word from v12, 1..3 words from v11
+    //    //  V12 = {Wt-8, Wt-7, Wt-6, Wt-5}
+    //    //  V11 = {Wt-12, Wt-11, Wt-10, Wt-9}
+    //    // Output
+    //    //  Vt4 = {Wt-12, Wt-7, Wt-6, Wt-5}
+    //    vmerge.vvm v14, v12, v11, v0
+    //
+    //    // Generate next Four Message Schedule Words (hence allowing for 4 more rounds)
+    //    // Input
+    //    //  V10 = {W[t+ 3], W[t+ 2], W[t+ 1], W[t+ 0]}     W[ 3: 0]
+    //    //  V13 = {W[t+15], W[t+14], W[t+13], W[t+12]}     W[15:12]
+    //    //  V14 = {W[t+11], W[t+10], W[t+ 9], W[t+ 4]}     W[11: 9,4]
+    //    // Output (next four message schedule words)
+    //    //  v10 = {W[t+19],  W[t+18],  W[t+17],  W[t+16]}  W[19:16]
+    //    vsha2ms.vv v10, v14, v13
+    //
+    // BEFORE
+    //  v10 - v13 hold the message schedule words (initially the block words)
+    //    v10 = W[ 3: 0]   "oldest"
+    //    v11 = W[ 7: 4]
+    //    v12 = W[11: 8]
+    //    v13 = W[15:12]   "newest"
+    //
+    //  vt6 - vt7 hold the working state variables
+    //    v16 = {a[t],b[t],e[t],f[t]}   // initially {H5,H4,H1,H0}
+    //    v17 = {c[t],d[t],g[t],h[t]}   // initially {H7,H6,H3,H2}
+    //
+    // AFTER
+    //  v10 - v13 hold the message schedule words (initially the block words)
+    //    v11 = W[ 7: 4]   "oldest"
+    //    v12 = W[11: 8]
+    //    v13 = W[15:12]
+    //    v10 = W[19:16]   "newest"
+    //
+    //  v16 and v17 hold the working state variables
+    //    v16 = {a[t+4],b[t+4],e[t+4],f[t+4]}
+    //    v17 = {c[t+4],d[t+4],g[t+4],h[t+4]}
+    //
+    //  The group of vectors v10,v11,v12,v13 is "rotated" by one in each quad-round,
+    //  hence the uses of those vectors rotate in each round, and we get back to the
+    //  initial configuration every 4 quad-rounds. We could avoid those changes at
+    //  the cost of moving those vectors at the end of each quad-rounds.
+
+    //--------------------------------------------------------------------------------
+    // Quad-round 0 (+0, v10->v11->v12->v13)
```

Comment:
similar comments as generate_sha256_implCompress about group the rounds in a loop.

Reply:
Now rounds share code.